ops/ccs-ops-misc/scripts/bfd-monthly-inventory

```diff
@@ -0,0 +1,288 @@
+  fi
+  if [[ "$br" =~ $br_name_re ]]; then
+    printf "Grabbing the latest from %s.. " "$br_name"
+    if git pull >/dev/null 2>&1; then
+      echo "OK"
+    else
+      echo "ERROR"
+      error_exit "could not pull from $br"
+    fi
+  else
+    error_exit "not on $br_name branch.. please 'git checkout $br_name' and try again."
+  fi
+}
+
+get_java_dependencies(){
+  echo
+  mkdir -p "$java_inv_dir"
+  cd "$APPS_DIR" || error_exit "could not access apps directory"
+  if $BUILD_APPS; then
+    printf "We need to build apps before getting dependencies (this may take a few minutes)... "
+    if mvn clean install -DskipITs >/dev/null 2>&1; then
+      echo "OK"
+    else
+      echo "ERROR"
+      error_exit "could not build apps"
+    fi
+  fi
+
+  printf "Exporting java dependencies.. "
+  if deps=$($mvn_dep_cmd | grep ":.*:.*:.*" | cut -d] -f2- | sed 's/:[a-z]*$//g' | sort -u); then
+    echo "OK"
+  else
+    error_exit "could not generate dependencies using '$mvn_dep_cmd'"
+  fi
+  echo "$deps" > "${java_inv_dir}/java-dependencies.txt"
+}
+
+# selects a server to inventory
+# $1 == environment (bfd-prod-sbx bfd-test prod-etl prod-sbx-etl test-etl)
+# returns an ip address
+select_target(){
+  # TODO: make this dynamic
+  PS3="Select an IP to inventory: "
+  # display a list of active ip's for the selected environment
+  select ip in $(aws ec2 describe-instances \
+  --filters 'Name=instance-state-name,Values=running' \
+  --query 'Reservations[].Instances[].[PrivateIpAddress,Tags[?Key==`Name`]| [0].Value]' \
+  --output table | grep "$1" | awk '{print $2}' | grep -v "None")
+  do
+    echo "$ip"
+    return
+  done
+}
+
+# runs a command on a remote ip using ssh
+# $1 == ip address of the target
+# $2 == cmd to run
+ssh_cmd(){
+  # 2>/dev/null to not print ssh banner
+  if ! ssh -i "$BFD_SSH_KEY_PATH" -t "${BFD_SSH_USER}@${1}" "$2" 2>/dev/null; then
+    echo "Failed to run '$2' on '$1'.. is the vpn up?"
+    error_exit "ssh command failure"
+  fi
+}
+
+# $1 == target to inventory
+# $2 == where to write results
+inventory_server(){
+  ssh_cmd "$1" "java -version" > "$2/java-dependencies.txt"
+  ssh_cmd "$1" "cat /etc/os-release" > "$2/os-release.txt"
+  ssh_cmd "$1" "sudo yum list installed --color=no" > "$2/yum-installed-packages.txt"
+  ssh_cmd "$1" "pip3 list" > "$2/python-packages.txt"
+}
+
+inventory_api_server(){
+  echo
+  mkdir -p "$api_inv_dir"
+  echo "We will now inventory one of the FHIR API servers from prod-sbx"
+  target="$(select_target "bfd-prod-sbx")"
+  printf "Inventorying %s.. " "$target"
+  inventory_server "$target" "$api_inv_dir"
+  echo "OK"
+}
+
+inventory_pipeline_server(){
+  echo
+  mkdir -p "$pipeline_inv_dir"
+  echo "We will now inventory one of the ETL/pipeline servers from prod-sbx"
+  target="$(select_target "bfd-prod-sbx-etl")"
+  printf "Inventorying %s.. " "$target"
+  inventory_server "$target" "$pipeline_inv_dir"
+  echo "OK"
+}
+
+
+#----------------- GO! ------------------#
```

Comment:
wondering if it should pre-clean here? every time that I ran it I forgot and got an error, then cleaned it up and ran it again.

Reply:
I now default to cleaning up on error, ctl-c, etc. Also added a CLEANUP env var to prevent this behavior (useful when debugging so you don't have to wait for the apps to compile every time).

The script should error out if the inventory dir is not empty on start.